### PR TITLE
Adding support for  'of collection' in Source Generator

### PR DESCRIPTION
**Pull request by kchinburarat**
Sync feature from reflection API to Source Generator

### Registration code
```csharp
using Agoda.IoC.Generator.Abstractions;

namespace GeneratorExample;

[RegisterScoped(Concrete = true)]
public class ClassA : IClassA
{
}
public interface IClassA { }


[RegisterSingleton(Factory = typeof(ClassBImplementationFactory))]
public class ClassC : IClassC
{
}
public interface IClassC { }


public class ClassBImplementationFactory : IImplementationFactory<IClassC>
{
    public IClassC Factory(IServiceProvider serviceProvider)
    {
        return new ClassC();
    }
}

[RegisterSingleton(For = typeof(IPipeline), OfCollection = true, Order = 2)]
public class Pipeline2 : IPipeline
{
    public string Invoke()
    {
        return nameof(Pipeline2);
    }
}
[RegisterSingleton(For = typeof(IPipeline), OfCollection = true, Order = 3)]
public class Pipeline3 : IPipeline
{
    public string Invoke()
    {
        return nameof(Pipeline2);
    }
}
[RegisterSingleton(For = typeof(IPipeline), OfCollection = true, Order = 1)]
public class Pipeline1 : IPipeline
{
    public string Invoke()
    {
        return nameof(Pipeline1);
    }
}
public interface IPipeline
{
    string Invoke();
}
[RegisterSingleton(For = typeof(IMiddleware), OfCollection = true, Order = 3)]
public class IMiddleware2 : IMiddleware
{
    public string Invoke()
    {
        return nameof(IMiddleware1);
    }
}
[RegisterSingleton(For = typeof(IMiddleware), OfCollection = true, Order = 1)]
public class IMiddleware1 : IMiddleware
{
    public string Invoke()
    {
        return nameof(IMiddleware1);
    }
}
public interface IMiddleware
{
    string Invoke();
}

```
### Generated Code 
```csharp
// <auto-generated>
//     Automatically generated by Agoda.IoC.Generator.
//     Changes made to this file may be lost and may cause undesirable behavior.
// </auto-generated>
using Microsoft.Extensions.DependencyInjection.Extensions;
using GeneratorExample;

namespace Microsoft.Extensions.DependencyInjection
{
    public static class AgodaIoCGeneratorInjectServiceCollectionExtension
    {
        public static IServiceCollection RegisterFromGeneratorExample(this IServiceCollection serviceCollection)
        {
            return Register(serviceCollection);
        }

        internal static IServiceCollection Register(this IServiceCollection serviceCollection)
        {
			serviceCollection.AddScoped<ClassA>();
			serviceCollection.AddSingleton(sp => new ClassBImplementationFactory().Factory(sp));
			// Of Collection code
			serviceCollection.AddSingleton<IMiddleware, IMiddleware1>();
			serviceCollection.AddSingleton<IMiddleware, IMiddleware2>();
			serviceCollection.AddSingleton<IPipeline, Pipeline1>();
			serviceCollection.AddSingleton<IPipeline, Pipeline2>();
			serviceCollection.AddSingleton<IPipeline, Pipeline3>();

            return serviceCollection;
        }
    }
}
```

### CODE REVIEW

**Comment by vladbatushkov** [comment]
Good to add a very simple test-case for Concrete = false.

**Review by vladbatushkov — APPROVED** [review]

File: src/Generator/Agoda.IoC.Generator/AgodaIoCGenerator.cs [hunk 1]
```diff
@@ -102,6 +128,22 @@ private static void Execute(Compilation compilation, ImmutableArray<ClassDeclara
...
 
+        AddSourceInternal(ctx, nsbuilder.ToString(), assemblyNameForMethod, registrationCodes.ToString());
+    }
+
+    private static void AddSourceInternal(SourceProductionContext ctx,
+        string usingNamespace,
+        string extensionMethodNamespace,
+        string registrationCodes)
+    {
+        var generatedCode = Constants.GENERATE_CLASS_SOURCE
```
Comment (szaboopeeter) [inline src/Generator/Agoda.IoC.Generator/AgodaIoCGenerator.cs]: `string.Format()` might be a bit more readable than 3 `Replace()`es.

File: src/Generator/Example/GeneratorExample/Exampler1.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
+﻿using Agoda.IoC.Generator.Abstractions;
+
```
Comment (szaboopeeter) [inline src/Generator/Example/GeneratorExample/Exampler1.cs]: I think you have a typo in the file name (Expample`r`1.cs)

Comment (kchinburarat) [reply src/Generator/Example/GeneratorExample/Exampler1.cs]: fixed

File: src/Generator/Agoda.IoC.Generator/AgodaIoCGenerator.cs (hunk 1 above)
Comment (kchinburarat) [reply src/Generator/Agoda.IoC.Generator/AgodaIoCGenerator.cs]: I try string.format but got `invalid input string` so maybe because of multi-line string

**Review by dicko2 — APPROVED** [review]

**Comment by kchinburarat** [comment]
> Good to add a very simple test-case for Concrete = false.

@vladbatushkov Already fixed code and adding more test cases
Thanks :)

**Pull request merged**